toolz/dicttoolz.py

```diff
@@ -48,14 +57,20 @@ def merge_with(func, *dicts):
     if len(dicts) == 1 and not isinstance(dicts[0], dict):
         dicts = dicts[0]
 
-    result = {}
+    dict_ = OrderedDict
+    result = OrderedDict()
+    return_ordered_dict = True
     for d in dicts:
+        if return_ordered_dict and not isinstance(d, OrderedDict):
+            result = dict(result)
+            dict_ = dict
+            return_ordered_dict = False
```

Comment:
My suggestion in that case is to actually make `dicts` concrete by calling `list` on it.  This would be bad if we benfitted by laziness but, due to the current implementation of `merge_with` we're not lazy anyway.